### DISCUSSION TITLE
Using Camera

### DISCUSSION THREAD

**Discussion by mizuki1412** [wailsapp/wails#883] [Q&A] [answered]
When using camera, I have to click the allow button every time, is there any way to avoid it? 
Something like "unsafely-treat-insecure-origin-as-secure" of chrome?

**Comment by leaanthony** [comment, marked as answer]
The best thing you can do is ask/search over at the webview2feedback repo and if they suggest that there's a webview2 preference that needs setting we can look at supporting it https://github.com/MicrosoftEdge/WebView2Feedback

Excited to see what you're building with the camera!

**Reply by mizuki1412** [reply]
thanks

**Reply by mizuki1412** [reply]
I find this https://github.com/MicrosoftEdge/WebView2Feedback/issues/742

Maybe can add some way to access the PermissionRequested event in the future ?

**Reply by leaanthony** [reply]
Brilliant, we can totally do that. Have a think about how you think the API would look. Maybe a callback in `options.windows`?

**Reply by leaanthony** [reply]
Also, do you have a test repo so I can test this? Thanks

**Reply by mizuki1412** [reply]
Yes, I think can add in the `windows.Options` .

And, I will prepare a test repo soon. Thanks!

**Reply by mizuki1412** [reply]
Hi, this is my test repo https://github.com/mizuki1412/webcam-test-wails. Thanks

**Reply by leaanthony** [reply]
Hi @mizuki1412 - please try latest version on `master`. Webcam working 👍

<img width="571" alt="2021-10-23 06_58_31-animate – main js" src="https://user-images.githubusercontent.com/1943904/138518085-8af85a2e-8ada-430e-9512-771c5e6318c4.png">

**Reply by mizuki1412** [reply]
nice!! It works well, Thanks!

**Comment by hermanbanken** [comment]
Wails 2.0 on WebKit doesn't seem to expose navigator.mediaDevices. Is this a known issue, or is this new in Wails 2.0?

**Reply by leaanthony** [reply]
We don't actively block it, but perhaps it needs user permissions? What does the WebKit documentation say?

**Reply by hermanbanken** [reply]
I see nothing on the topic of cameras listed:
https://developer.apple.com/documentation/webkit/wkwebviewconfiguration

**Reply by hermanbanken** [reply]
Same boat as this person: https://stackoverflow.com/questions/71333432/how-can-i-know-if-a-web-api-is-supported-by-wkwebview

**Reply by hermanbanken** [reply]
Here in these WWDC notes it is mentioned that getUserMedia is added to WKWebView:
https://www.wwdcnotes.com/notes/wwdc21/10032/
> Control media capture (getUserMedia)
>
> WKWebView supports getUserMedia from iOS 14.3; which allows WebRTC functions to work inside your app

for this to work, they mention implementing a method that decides whether to prompt or deny the request:

```swift
func webView(
    _ webView: WKWebView,
    decideMediaCapturePermissionFor origin: WKSecurityOrigin,
    initiatedByFrame frame: WKFrameInfo,
    type: WKMediaCaptureType
  ) async -> WKPermissionDecision {
    return type == .microphone? .prompt: .deny
  }
```

**Reply by hermanbanken** [reply]
This doc notes 
> If you don’t implement this method in your delegate, the system returns [WKPermissionDecision.prompt](https://developer.apple.com/documentation/webkit/wkpermissiondecision/prompt).

so either Wails is actively denying, or something weird is going on here.

**Reply by leaanthony** [reply]
Maybe it's denied by default. If there isn't a ticket for this, please open one and we'll add it to the backlog 👍

**Reply by hermanbanken** [reply]
https://github.com/wailsapp/wails/blob/b93cdcd942cc70686fc2790038707496948ff7ea/v2/internal/frontend/desktop/darwin/WailsContext.m

Nothing here, but this seems fairly trivial. I’ll add an issue/PR tomorrow (EU-based).

**Reply by marcio199226** [reply]
any progress about this ?

**Reply by leaanthony** [reply]
No. Do you fancy taking it on? Happy to accept a PR!

**Reply by iamyellow** [reply]
old issue, but i'm able to PR this since i'm also developing an app that uses media streams. i'll try today

**Comment by fairking** [comment]
I am also thinking of camera implementation. Just wondering would it be better to capture camera in backend and then stream the video to the frontend? Also to control the camera by sending commands from the frontend? Why do we need to handle it by the webview?

As far as I know the byte streaming is not supported but it could open a wide range of possibilities like video calls, music player, qr code scanner etc.

**Reply by leaanthony** [reply]
I managed to make a webcam work fine on Windows from frontend using a media element. It was 2y ago so I can't quite remember the details. Search the issue tracker for 'media'.

**Reply by fairking** [reply]
I am not interested in frontend (webview) camera integration. I would like to process some camera data in backend using golang and then stream the result to the frontend. 

Something like that: [CreateSharedBuffer](https://github.com/MicrosoftEdge/WebView2Feedback/blob/main/specs/SharedBuffer.md). It is not actually a stream but I guess it could send such buffers as chunks one by one simulating a stream.

Apart of that I think any hardware integration (sensors) should happen in backend. The WebView is a representation layer and it should not be used to process any data (nor integrate hardware), but to be able to receive such data for representation (rendering).

Please note: Wails role is not for camera (or any other hardware) integration support. Wails should provide a solid and simple communication support between the presentation layer (WebView) and the business layer (backend, golang). Everything else is up to devs what they can do with that. IMHO.

Examples:
- https://github.com/vee2xx/camtron (removed)
- https://github.com/AlexxIT/go2rtc
- https://medium.com/go4vl/building-a-webcam-with-go-and-go4vl-7b56d2c54e39
- https://medium.com/learning-the-go-programming-language/realtime-video-capture-with-go-65a8ac3a57da

**Reply by zhaojunliing** [reply]
camtron 

It uses electron to get the camera and send it to the backend.

**Reply by fairking** [reply]
So in this case it is not suitable to me then. And again, it is not about camera, but about ability to stream data to the frontend (webview).

**Reply by leaanthony** [reply]
We'll look at supporting CreateSharedBuffer in v3 👍 I was looking at the docs yesterday.

**Reply by iamkhalidbashir** [reply]
Has v3 planned for this?

**Comment by madeindra** [comment]
My Wails apps need microphone access and days ago it work perfectly normal without any config, today it can't get access to `navigator.mediaDevices` saying it's undefined.

I'm on MacOS, seems like adding this to `Info.plist` and `Info.dev.plist` between `<dict> </dict>` solves it

```
<key>NSMicrophoneUsageDescription</key>
<string>This app requires access to the microphone</string>
```

If you need camera access in your wails app and it keep saying `navigator.mediaDevices` is undefined, try using this:
```
<key>NSCameraUsageDescription</key>
<string>This app requires access to the camera</string>
```

**Reply by ndelvalle** [reply]
I wonder if there is a way to setup this without modifying the build folder, something that can generate this for every build target.
